operator/src/main/java/org/bf2/operator/managers/CapacityManager.java

```diff
@@ -0,0 +1,330 @@
+                OperandUtils.getOrDefault(mk.getMetadata().getLabels(), ManagedKafka.PROFILE_QUOTA_CONSUMED, "1"));
+        return resources;
+    }
+
+    public synchronized void releaseResources(ManagedKafka managedKafka) {
+        // use the latest
+        ConfigMap resourceConfigMap = client.configMaps().withName(FLEETSHARD_RESOURCES).get();
+        if (resourceConfigMap != null) {
+            String entryKey = getManagedKafkaKey(managedKafka);
+            if (releaseResources(resourceConfigMap.getData(), entryKey)) {
+                client.configMaps()
+                        .withName(FLEETSHARD_RESOURCES)
+                        .lockResourceVersion(resourceConfigMap.getMetadata().getResourceVersion())
+                        .replace(resourceConfigMap);
+            }
+        }
+    }
+
+    static String getManagedKafkaKey(ManagedKafka managedKafka) {
+        return "mk-" + managedKafka.getId();
+    }
+
+    private boolean releaseResources(Map<String, String> resourceData, String entryKey) {
+        String previous = resourceData.remove(entryKey);
+        if (previous != null) {
+            Resources resources = Serialization.unmarshal(previous, Resources.class);
+            resourceData.computeIfPresent(resources.profile,
+                    (k, v) -> String.valueOf(Integer.parseInt(v) - resources.units));
+            return true;
+        }
+        return false;
+    }
+
+    public synchronized Optional<OperandReadiness> claimResources(ManagedKafka managedKafka, String profile,
+            ManagedKafkaAgent agent) {
+        Integer max = getProfileMaxUnits(agent.getSpec().getCapacity(), profile).orElse(null);
+        if (max == null) {
+            return Optional.empty();
+        }
+
+        // ensure we have the latest
+        ConfigMap resourceConfigMap = client.configMaps().withName(FLEETSHARD_RESOURCES).get();
+
+        Resources resources = createResources(managedKafka, profile);
+
+        int total = Integer.parseInt(resourceConfigMap.getData().getOrDefault(profile, "0"));
+
+        int newTotal = resources.units + total;
+        if (newTotal > max) {
+            log.warnf("Rejecting the deployment of %s/%s as the cluster is full",
+                    managedKafka.getMetadata().getNamespace(), managedKafka.getMetadata().getName());
+            return Optional
+                    .of(new OperandReadiness(Status.False, Reason.Rejected, "Cluster has insufficient resources"));
+        }
+
+        // update the configmap in a locked manner - this may fail, but we'll retry
+        resourceConfigMap.getData().put(profile, String.valueOf(newTotal));
+        resourceConfigMap.getData().put(getManagedKafkaKey(managedKafka), Serialization.asJson(resources));
+        client.configMaps()
+                .withName(FLEETSHARD_RESOURCES)
+                .lockResourceVersion(resourceConfigMap.getMetadata().getResourceVersion())
+                .replace(resourceConfigMap);
```

Comment:
If it does fail, is it a runtime exception? I'm wondering if it's something that should be handled that's expected to occur somewhat frequently.

Reply:
It would fail with a RuntimeException - KubernetesClientException.  It's not expect to occur frequently.  It would be a corner case of a manual modification or another operator pod making a modification.